addons/api/addon/serializers/host-catalog.js

```diff
@@ -4,115 +4,43 @@
     const serialized = super.serialize(...arguments);
-
-    // This is a hack to remove the worker_filter field from the serialized data if the compositeType is not AWS
-    // Ideally we should refactor the model to include the compositeType using `for` attribute
-    if (compositeType !== TYPE_HOST_CATALOG_PLUGIN_AWS) {
-      delete serialized.worker_filter;
-    }
-
     // By default, disable credential rotation for dynamic credentials
     if (credentialType === TYPE_CREDENTIAL_DYNAMIC) {
       serialized.attributes.disable_credential_rotation = true;
     }
-
-    switch (compositeType) {
-      case 'static':
-        return this.serializeStatic(...arguments);
-      default:
-        return serialized;
-    }
+    return serialized;
   }
 
   serializeAttribute(snapshot, json, key, attribute) {
     const value = super.serializeAttribute(...arguments);
-    const { compositeType, credentialType, isPlugin } = snapshot.record;
+    const { compositeType, credentialType } = snapshot.record;
     const { options } = attribute;
-    const fields =
-      compositeType === TYPE_HOST_CATALOG_PLUGIN_AZURE
-        ? azureFields
-        : AWSfieldsWithCredentialType[credentialType];
-
-    // Delete any nested attribute fields that don't belong to the record's compositeType
-    if (isPlugin && options.isNestedAttribute && json.attributes) {
-      // If compositeType is AWS, delete azure fields from the json, and set the rest that don't belong to respective compositeType to null
-      if (compositeType === TYPE_HOST_CATALOG_PLUGIN_AWS) {
-        if (!AWSfieldsWithCredentialType[credentialType].includes(key)) {
-          if (azureFields.includes(key)) {
-            delete json.attributes[key];
-          } else {
-            json.attributes[key] = null;
-          }
-        }
-      }
-
-      // If compositeType is Azure, delete AWS fields from the json
-      if (
-        compositeType === TYPE_HOST_CATALOG_PLUGIN_AZURE &&
-        !azureFields.includes(key)
-      ) {
-        delete json.attributes[key];
-      }
-    }
-
-    // Delete any secret fields that don't belong to the record type
-    if (isPlugin && options.isNestedSecret && json.secrets) {
-      if (!fields.includes(key)) {
+    const isSamePluginName = options?.for?.name?.includes(compositeType);
+
+    // Remove attributes that are not applicable to the current credential type for the same plugin
+    // For example, 'aws' dynamic plugin has two credential types: 'static' and 'dynamic'
+    // API expects the fields to be 'null' for the fields that are not from the current credential type (that's why we are not using the application serializer for this, where we delete attrs)
+    if (
+      isSamePluginName &&
+      typeOf(options.for) === 'object' &&
+      options.for.credentialType &&
+      options.for.credentialType !== credentialType
+    ) {
+      if (options.isNestedAttribute) {
+        json.attributes[key] = null;
```

Comment:
How come we have to set it to `null` if it's an attribute but delete it if it's a nested secret or a normal top level field?

Reply:
this was based on a discussion from the backend team that we need to set these to null, but I'm going to double check about secrets and the top level fields

Reply:
this is expected when the user makes an update from dynamic to static `credentialType`, we have to explicitly pass null for assume role fields which are nested attributes. But the secrets and top-level fields follow the usual pattern where we delete the fields that don't belong to the type. I asked for clarification here in [slack](https://hashicorp.slack.com/archives/C07E8C15JGL/p1735936709962049)